New function to add  Foods to the MFP DB

Dnic94 opened the pull request:
Hey there,

I hope this finds you well.

I have added a function to submit new food to the MFP database. I tried to respect your code style and use already implemented functions, logging and exception handling.

Feel free to merge or contact me with any questions or comments.

Greetings 
Dominic

neilramaswamy commented on myfitnesspal/client.py:
```diff
@@ -795,3 +795,127 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                    # print("No Error :)")
+                    # Would like to return FoodItem, but seems that it take
+                    # to long until the submitted food is available in the DB
+                    # return self.get_food_search_results("{} {}".format(brand, description))[0]
```
Yeah, hmm, this looks tough. In some ways it's not entirely our job to fix their read-after-write inconsistency, so I'm not sure we should try to hack it on our end. And IMO, we probably don't need to return a `FoodItem`, since within the web application, the UX doesn't seem to be showing the created item after it is created—it just redirects me back to my diary. So we can probably just `return` here without worrying about the get food search results call.

(But we can and should defer to the maintainers. This is just my take.)

Dnic94 replied:
I agree with you. Was just an idea for a nice to have feature :)

neilramaswamy commented on myfitnesspal/client.py:
```diff
@@ -795,3 +795,127 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                    # return self.get_food_search_results("{} {}".format(brand, description))[0]
+                else:  # Error occurred
+                    error = document.xpath("//*[@id='errorExplanation']/ul/li")[0].text
+                    error = error.replace("Description ", "")
```
Why do we need this? I took a look at `#errorExplanation` and didn't see any text of "Description ".

Dnic94 replied:
Just for cosmetic reasons.  At least I got some errors where "Description " was placed before the error message. Just looked odd in the log/print out,

hannahburkhardt replied:
I also get errors that start with "Description " e.g. 
`'Description An item with this brand and description already exists.'`
so this is a fine replacement to make I think.

hannahburkhardt reviewed (commented):
Hi Dominic, 
you have a fair number of whitespace changes in existing parts of this file. Did you mean to make those changes?
For the doc string, it might be good to clarify that this is adding new food entries to the MFP DB (as in the PR title) to avoid confusion w/ logging a diary item. It might also be good for the docstring to mention the return value and its meaning.
The trans_fat argument type definition has a typo, I think; as does servingspercontainer default value.
On line 911 and 914 there is no placeholder for the error so the error will not be printed.

neilramaswamy commented:
I also just realized that we probably should add this to the documentation/README. I'd be happy to do so if you'd like, just let me know.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -795,3 +795,127 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+    ### Dominic Schwarz (Dnic94) <dominic.schwarz@dnic42.de> - 25.08.2021 ###
+    ### Added function to submit new foods to MFP.
+    SUBMIT_PATH = "food/submit"
+    SUBMIT_DUPLICATE_PATH = "food/duplicate"
```
In the majority (~9 out of 10) of test cases I was redirected to food/duplicate after post the data to food/new. So i had to do post it again. Posting it directly to food/duplicate just worked fine for me everytime.

neilramaswamy replied:
What about posting to `food/submit`?

Dnic94 replied:
I could not observe a post to food/submit and therefore not tested it. I just reversed the way how it works in the browser and implemented it like that. To be honest i don't see an issue here, but I am open for a better solution if you observed or found one.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -795,3 +795,127 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        # Step 3 - Passed Brand and Desc. Ready submit Form but needs new Authenticity Token
+        submit3_url = parse.urljoin(self.BASE_URL_SECURE, self.SUBMIT_NEW_PATH)
+        document = self._get_document_for_url(submit3_url)
+        authenticity_token = document.xpath(
```
Interesting. Did not work for me with reusing a token. Also discovered in the Web-Developer Console of my browser that the token changed after the request. And if i remember correctly I saw in the code that the token will also read out again at the add measurement functions before posting data.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -795,3 +795,127 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+
+        # Step4 - Build Post Data and finally submit new Food with nutritional Details
+        data = {
+            "utf8": utf8_field,
```
Could be the case. Did not checked it to be honest. Just oriented at the add measurement code.

Dnic94 commented:
> Hi Dominic,
> you have a fair number of whitespace changes in existing parts of this file. Did you mean to make those changes?
> For the doc string, it might be good to clarify that this is adding new food entries to the MFP DB (as in the PR title) to avoid confusion w/ logging a diary item. It might also be good for the docstring to mention the return value and its meaning.
> The trans_fat argument type definition has a typo, I think; as does servingspercontainer default value.
> On line 911 and 914 there is no placeholder for the error so the error will not be printed.

Hey Hannah,
You are right. Fixed all the addressed issues, but whats the problem with 1.0 as default for servingspercontainer? 
The whitespace changes came from PyCharms reformat code on commit feature. 

Thanks for the review.

hannahburkhardt commented:
`servingspercontainer: float = "1.0",`
You specify the datatype as float but the default value is a string :)

Dnic94 commented:
> `servingspercontainer: float = "1.0",`
> You specify the datatype as float but the default value is a string :)
> In fact you also specify the other numbers to be of float type but then supply a string type default value. The type definition in python is a suggestion that is not enforced, but this wouldn't work in a strictly typed language... You might consider using None as default value, and checking for None when formatting, but since it doesn't really matter in python, I would consider that a style issue.
> When in doubt it's best to follow the existing style. From set_measurements, that would be specifying None as default value and formatting your post data without explicitly converting to a string, like so:
> `"measurement[display_value]": value,`

Just too obvious. 🙃  But for sure you are right. None as default value would be the correct way in my opinion but would also lead to a "bad" data-block for posting without previous checking and replacing None with "" again. I am not that familiar with all the format and styling options. I will follow your suggestions and have another look at the method which was used in set_measurements. Otherwise a sub-function like

```
def nonetoemptystr(value):
    if type(value) == None:
        return ""
    else:
        return value
```

should also work.
Thanks again and have a nice day!

Dnic94 commented:
He,

I added some additional functions for

- Adjusting your nutrition goals
- Gather your meals
- Gather your recipes
- Gather meal details parsed into recipe schema from schema.org
- Gather recipe details parsed into recipe schema from schema.org

hannahburkhardt commented:
Hi @Dnic94, since this is a pretty large change, I gave it a thorough-ish code style review. I hope you find my comments helpful and not just annoying...! Thanks for your contribution!

Dnic94 commented:
> Hi @Dnic94, since this is a pretty large change, I gave it a thorough-ish code style review. I hope you find my comments helpful and not just annoying...! Thanks for your contribution!

Hey @hannahburkhardt, please feel free to leave your comments. 
Your feedback is always appreciated.  :)

Dnic94 commented on .gitignore:
```diff
@@ -15,3 +15,5 @@ pip-selfcheck.json
 env/*
 .vscode/*
 .mypy_cache/*
+test_file.py
```
Nope, it is just a private test file. Will remove it with the next commit.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+
+        # get Authenticity Token
+        url = parse.urljoin(self.BASE_URL_SECURE, SUBMIT_PATH)
+        # now = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
```
Done.  Will be removed with the next commit.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        data = {
+            "utf8": utf8_field,
+            "authenticity_token": authenticity_token,
+            "date": datetime.datetime.today().strftime("%Y-%m-%d"),
```
Hmm, sounds reasonable. Will do so :)

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        recipes_dict = {}
+
+        page_count = 1
+        next_page = True
```
Done. Will be renamed in the next commit.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            ].text
+
+            recipe_dict["recipeIngredient"] = []
+            ingridients = document.xpath('//*[@id="main"]/div[4]/div/*/li')
```
Hehe, sorry for that.  😅

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            "food[brand]": brand,
+            "food[description]": description,
+            "weight[serving_size]": serving_size,
+            "servingspercontainer": "{}".format(servingspercontainer),
```
You are right. Also switched my styling to f-strings since i wrote this code.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        percent_carbohydrates: float = "",
+        percent_protein: float = "",
+        percent_fat: float = "",
+        # saturated_fat: float = "",
+        # polyunsaturated_fat: float = "",
+        # monounsaturated_fat: float = "",
+        # trans_fat: float = "",
+        # fiber: float = "",
+        # sugar: float = "",
+        # cholesterol: float = "",
+        # sodium: float = "",
+        # potassium: float = "",
+        # vitamin_a: float = "",
+        # vitamin_c: float = "",
+        # calcium: float = "",
+        # iron: float = "",
```
IMO we can leave it there. It will not hurt and if someone wants to implement the micro nutritions it is already prepared.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+
+        # Get authenticity token and current values
+        url = parse.urljoin(self.BASE_URL_SECURE, "account/my_goals")
+        # now = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
```
Done.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        # now = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
+        today = datetime.datetime.now().strftime("%Y-%m-%d")
+
+        """
+        old_goals_url = parse.urljoin(self.BASE_API_URL, f"v2/nutrient-goals?date={today}")
+        #No Way implemented to send token by calling self._get_json_for_url
+        old_goals = self._get_json_for_url(old_goals_url)
+        """
```
Yes, will leave it there as explanation why we do not use self._get_json_for_url(old_goals_url) here at the moment.

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,545 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        fat: float = "",
+        percent_carbohydrates: float = "",
+        percent_protein: float = "",
+        percent_fat: float = "",
```
As the the new values will be taken from the old values in case they are not provided it does not make sense here to assign an empty string here. 

I did it because we observed that MFP responded with an error on our post if we sent a None in the payload instead of "".

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        percent_carbohydrates: float = None,
+        percent_protein: float = None,
+        percent_fat: float = None,
+        # TODO Add micro nutritions
+        # saturated_fat: float = None,
+        # polyunsaturated_fat: float = None,
+        # monounsaturated_fat: float = None,
+        # trans_fat: float = None,
+        # fiber: float = None,
+        # sugar: float = None,
+        # cholesterol: float = None,
+        # sodium: float = None,
+        # potassium: float = None,
+        # vitamin_a: float = None,
+        # vitamin_c: float = None,
+        # calcium: float = None,
+        # iron: float = None,
```
```suggestion
```

I'd really rather keep the code as clean as possible; so if you wouldn't mind, let's either remove these comments or actually implement them.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        url = parse.urljoin(self.BASE_URL_SECURE, "account/my_goals")
+        today = datetime.datetime.now().strftime("%Y-%m-%d")
+
+        """
+        Will leave this here as reminder why we do not use self._get_json_for_url() at the moment.
+        # No Way implemented to send token by calling self._get_json_for_url
+        old_goals_url = parse.urljoin(self.BASE_API_URL, f"v2/nutrient-goals?date={today}")
+        old_goals = self._get_json_for_url(old_goals_url)
+        """
```
```suggestion
        # Will leave this here as reminder why we do not use self._get_json_for_url() at the moment.
        # No Way implemented to send token by calling self._get_json_for_url
        # old_goals_url = parse.urljoin(self.BASE_API_URL, f"v2/nutrient-goals?date={today}")
        # old_goals = self._get_json_for_url(old_goals_url)
```

I'm not quite sure I understand what the comment is explaining, though -- I wonder if you would mind elaborating on what it's trying to remind future-us about?

Dnic94 replied:
Sure. to get a proper response from MFP we need to send additional headers with our post to the BASE_API_URL.
But it is not implemented in _get_json_for_url() to pass these to the function.

But in the end it does not really matter as it would only save 1 or 2 lines. I will remove it to keep the code clean.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                or percent_fat is None
+            ):
+                if energy is None:
+                    logger.warning("No energy value and no macro values provided! - Not able to update goals without a mandotory amount of information.")
```
```suggestion
                    raise ValueError("No energy value and no macro values provided! - Not able to update goals without a mandatory amount of information.")
```

This, too, seems like a situation where we should be raising an exception.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                    # If old and new values are in diffrent units then convert old value to new unit
+                    if not old_energy_unit == energy_unit:
+                        if old_energy_unit == "kilojoules" and energy_unit == "calories":
+                            old_energy_value *= 0.2388
```
I'm sure it's not going to matter much, but I _believe_ the actual value is 0.239006, no?

Dnic94 replied:
I took the values from the MFP javascript to match the original behaviour. :)

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                            old_energy_value *= 4.184
+                            old_energy_unit = "kilojoules"
+                        else:
+                            raise ValueError
```
We could probably be a little more helpful in the exception message here, and probably `ValueError` isn't the right choice since this problem isn't occurring due to a value that the user supplied, and is instead occurring due to the shape of their historical goals not meeting one of the above cases.

That all being said -- I kind of wonder why we'd bother with trying to convert the units at all here; is there a reason we need to do that?

Dnic94 replied:
I'll look at this again.

Dnic94 replied:
I think it make sense to do this conversion, otherwise the calculation afterwards would result in garbage if the user provided arguments and the MFP user preferences mismatch.

So I adjusted the exception message.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            # If percentage values were provided check
+            else:
+                if energy is None:
+                    logger.warning("No energy value and no macro values provided! - Not able to update goals without a mandotory amount of information.")
```
This is also a situation in which we should be raising an exception instead of just logging.
```suggestion
                    raise ValueError("No energy value and no macro values provided! - Not able to update goals without a mandotory amount of information.")
```

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                            protein = round(protein / 4.184, 2)
+                            fat = round(fat / 4.184, 2)
+                    else:
+                        logger.warning("Provided percentage values are not plausible - Not able to update goals.")
```
& here, too, I think, we should raise an exception.
```suggestion
                        raise ValueError("Provided percentage values are not plausible - Not able to update goals.")
```

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+
+        # Marcro Calculation
+        # If no macro goals were provided calculate them with percentage value
+        if carbohydrates is None or protein is None or fat is None:
```
I'm sort of of the mind right now that maybe we shouldn't try to handle missing information at all; it'd make the logic a lot simpler below, and I wonder what use case we're imagining for somebody to only provide a subset of these?

Dnic94 replied:
The idea was that you can pass absolut values or percentage values. I think there are use cases for both.
But i am with you that the logic looks really complex. Do you think it would be better to split it into two functions?

Dnic94 replied:
I added an explaining doc string to the function. How it works should be clear now.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                energy = macro_energy
+
+        # Build payload based on observed browser behaviour
+        # TODO Insert additional micro nurtitions
```
Could I convince you into removing these TODOs?  I understand that you're trying to guide future writers, but very often comments like this end up just becoming noise.

Dnic94 replied:
Done.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        result = self.session.post(url, json.dumps(new_goals), headers=auth_header)
+
+        # TODO Check Request Result
+        if result.status_code == 200:
+            return True
+        elif not result.ok:
+            logger.warning(
+                "Request Error - Unable to submit Goals to MyFitnessPal: "
+                "status code: {status}".format(status=result.status_code)
+            )
+            return False
+        else:
+            logger.error(
+                "Request Error - Unable to submit Goals to MyFitnessPal: "
+                "status code: {status}".format(status=result.status_code)
+            )
+            print(result)
+            return False
```
```suggestion
        if not result.ok:
            raise SomeKindOfException(
                "Request Error - Unable to submit Goals to MyFitnessPal: "
                "status code: {status}".format(status=result.status_code)
            )
```

I think the above is pretty self-explanatory, but in case it isn't, I believe this is functionally essentially the same as your original implementation in all ways except how it responds to the non-OK status (by raising an exception) & doesn't return a `True` value.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        )
+        if result.status_code == 200:
+            document = lxml.html.document_fromstring(result.content.decode("utf-8"))
+            try:
+                if not document.xpath(
+                    # If list is empty there should be no error, could be replaced with assert
+                    "//*[@id='errorExplanation']/ul/li"
+                ):
+                    return True
+                    # print("No Error :)")
+                    # Would like to return FoodItem, but seems that it take
+                    # to long until the submitted food is available in the DB
+                    # return self.get_food_search_results("{} {}".format(brand, description))[0]
+                else:  # Error occurred
+                    error = document.xpath("//*[@id='errorExplanation']/ul/li")[0].text
+                    error = error.replace("Description ", "")  # For cosmetic reasons
+                    raise MyfitnesspalRequestFailed(
+                        f"Unable to submit food to MyFitnessPal: {error}"
+                    )
+            except Exception:
+                logger.warning(f"Unable to submit food to MyFitnessPal: {error}")
```
```suggestion
            if not document.xpath(
                # If list is empty there should be no error, could be replaced with assert
                "//*[@id='errorExplanation']/ul/li"
            ):
                return True
                # print("No Error :)")
                # Would like to return FoodItem, but seems that it take
                # to long until the submitted food is available in the DB
                # return self.get_food_search_results("{} {}".format(brand, description))[0]
            else:  # Error occurred
                error = document.xpath("//*[@id='errorExplanation']/ul/li")[0].text
                error = error.replace("Description ", "")  # For cosmetic reasons
                raise MyfitnesspalRequestFailed(
                    f"Unable to submit food to MyFitnessPal: {error}"
                )
```

I kind of think we should just leave that exception uncaught so the caller can deal with it.

Dnic94 replied:
ok, added with the next commit.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                logger.warning(f"Unable to submit food to MyFitnessPal: {error}")
+
+        elif not result.ok:
+            logger.warning(
+                f"Request Error - Unable to submit food to MyFitnessPal: status code: {result.status_code}"
+            )
+            return None
```
```suggestion
            raise MyfitnesspalRequestFailed(
                f"Request Error - Unable to submit food to MyFitnessPal: status code: {result.status_code}"
            )
```

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        # calcium: float = None,
+        # iron: float = None,
+        assign_exercise_energy="nutrient_goal",
+    ):
```
A return type here would be helpful, too, if you could.

Dnic94 replied:
WIll be added with the next commits.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        old_goals_url = parse.urljoin(
+            self.BASE_API_URL, f"v2/nutrient-goals?date={today}"
+        )
+        old_goals_document = self.session.get(old_goals_url, headers=auth_header)
+        old_goals = json.loads(old_goals_document.text)
```
I'm kind of curious what this does if you have never set goals; isn't that an optional thing in the UI?  Maybe we should also be checking the status code of `old_goals_document` to make sure it's ok?

We could just do something like this, if I remember correctly:

```suggestion
        old_goals_document = self.session.get(old_goals_url, headers=auth_header)
        old_goals_document.raise_for_status()
        old_goals = json.loads(old_goals_document.text)
```

Dnic94 replied:
I will check this, but I think there should always be some default values available, even with a complete new account.

Dnic94 replied:
Checked it. You are forced to set a goal when you create your account. So there will be old goals available in any case.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+
+        return recipes_dict
+
+    def get_recipe(self, recipeid: int):
```
Would you mind declaring a return type here, too?  Especially for methods that return values, they help a lot because they show folks what properties your returned recipe might have for them to use.

Dnic94 replied:
Will be added with the next commits.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            print(result)
+            return False
+
+    def get_recipe_list(self):
```
:pray: for a return type, too, if you could.  Otherwise, folks can only know what this is going to return by trying it experimentally.

Also, looking at what you're actually returning below, you may want to either change the return type or rename this method -- you're returning a dictionary below rather than a list.

Dnic94 replied:
I think i will just rename the functions to get_recipes and get_meals. The list was not meant to be the return type.
It was just a hint that the function returns an overview of your recipes/meals and not the details of one recipe/meal.
So it was just to distinguish the functions.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                # Indicator for no recipes
+                has_next_page = False
+                if len(recipes_dict) == 0:
+                    return None
```
Probably should just return an empty dictionary so the caller only has to deal with a single response type, no?

Dnic94 replied:
I will change it :)

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        return recipes_dict
+
+    def get_recipe(self, recipeid: int):
+        recipe_PATH = f"/recipe/view/{recipeid}"
+        recipe_url = parse.urljoin(self.BASE_URL_SECURE, recipe_PATH)
```
Forgive me, this is a very nit-y request, but the standard python style would rather have this variable named recipe_path (here & below).

```suggestion
        recipe_path = f"/recipe/view/{recipeid}"
        recipe_url = parse.urljoin(self.BASE_URL_SECURE, recipe_path)
```

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            recipe_dict["nutrition"] = {"@type": "NutritionInformation"}
+            recipe_dict["nutrition"]["calories"] = document.xpath(
+                '//*[@id="main"]/div[3]/div[2]/div[2]/div'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["carbohydrateContent"] = document.xpath(
+                '//*[@id="carbs"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["fiberContent"] = document.xpath(
+                '//*[@id="fiber"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["sugarContent"] = document.xpath(
+                '//*[@id="sugar"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["sodiumContent"] = document.xpath(
+                '//*[@id="sodium"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["proteinContent"] = document.xpath(
+                '//*[@id="protein"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["fatContent"] = document.xpath(
+                '//*[@id="total_fat"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["saturatedFatContent"] = document.xpath(
+                '//*[@id="saturated_fat"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["monounsaturatedFatContent"] = document.xpath(
+                '//*[@id="monounsaturated_fat"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["polyunsaturatedFatContent"] = document.xpath(
+                '//*[@id="polyunsaturated_fat"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+            recipe_dict["nutrition"]["unsaturatedFatContent"] = int(
+                recipe_dict["nutrition"]["polyunsaturatedFatContent"]
+            ) + int(recipe_dict["nutrition"]["monounsaturatedFatContent"])
+            recipe_dict["nutrition"]["transFatContent"] = document.xpath(
+                '//*[@id="trans_fat"]/td[1]/span[2]'
+            )[0].text.strip(" \n")
+        except Exception:
+            logger.warning(f"Could not extract recipe information from {recipe_url}")
+            return None
```
```suggestion
        recipe_dict["name"] = document.xpath('//*[@id="main"]/div[3]/div[2]/h1')[
            0
        ].text
        recipe_dict["recipeYield"] = document.xpath('//*[@id="recipe_servings"]')[
            0
        ].text

        recipe_dict["recipeIngredient"] = []
        ingredients = document.xpath('//*[@id="main"]/div[4]/div/*/li')
        for ingredient in ingredients:
            recipe_dict["recipeIngredient"].append(ingredient.text.strip(" \n"))

        recipe_dict["nutrition"] = {"@type": "NutritionInformation"}
        recipe_dict["nutrition"]["calories"] = document.xpath(
            '//*[@id="main"]/div[3]/div[2]/div[2]/div'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["carbohydrateContent"] = document.xpath(
            '//*[@id="carbs"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["fiberContent"] = document.xpath(
            '//*[@id="fiber"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["sugarContent"] = document.xpath(
            '//*[@id="sugar"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["sodiumContent"] = document.xpath(
            '//*[@id="sodium"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["proteinContent"] = document.xpath(
            '//*[@id="protein"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["fatContent"] = document.xpath(
            '//*[@id="total_fat"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["saturatedFatContent"] = document.xpath(
            '//*[@id="saturated_fat"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["monounsaturatedFatContent"] = document.xpath(
            '//*[@id="monounsaturated_fat"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["polyunsaturatedFatContent"] = document.xpath(
            '//*[@id="polyunsaturated_fat"]/td[1]/span[2]'
        )[0].text.strip(" \n")
        recipe_dict["nutrition"]["unsaturatedFatContent"] = int(
            recipe_dict["nutrition"]["polyunsaturatedFatContent"]
        ) + int(recipe_dict["nutrition"]["monounsaturatedFatContent"])
        recipe_dict["nutrition"]["transFatContent"] = document.xpath(
            '//*[@id="trans_fat"]/td[1]/span[2]'
        )[0].text.strip(" \n")
```

We should probably let this exception go uncaught or the user won't have any way of telling what's going wrong for even submitting a bug report.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        recipe_dict["tags"] = ["MyFitnessPal"]
+        return recipe_dict
+
+    def get_meal_list(self):
```
I'm sure you know what I'm going to ask for here :-)

Dnic94 replied:
Will be added with the next commits.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        return meals_dict
+
+    def get_meal(self, mealid: int, meal_title: str):
+
```
```suggestion
```

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+                and ingredients[0].xpath("./td[1]")[0].text == "\xa0"
+            ):
+                raise
+            else:
+                for ingredient in ingredients:
+                    recipe_dict["recipeIngredient"].append(
+                        ingredient.xpath("./td[1]")[0].text
+                    )
+                    """ #Unfortunately the schema does not account for this information based on ingredients
+                    tmp = {}
+                    tmp['title'] = ingredient.xpath('./td[1]')[0].text
+                    tmp['nutrition'] = {}
+                    tmp['nutrition']['energy'] = ingredient.xpath('./td[2]')[0].text
+                    tmp['nutrition']['carbohydrates'] = ingredient.xpath('./td[3]')[0].text
+                    tmp['nutrition']['protein'] = ingredient.xpath('./td[5]')[0].text
+                    tmp['nutrition']['fat'] = ingredient.xpath('./td[4]')[0].text
+                    tmp['nutrition']['sugar'] = ingredient.xpath('./td[7]')[0].text
+                    tmp['nutrition']['sodium'] = ingredient.xpath('./td[6]')[0].text
+                    meal_dict['ingredients'].append(tmp)
+                    """
+
+                total = document.xpath('//*[@id="mealTableTotal"]/tbody/tr')[0]
+                recipe_dict["nutrition"] = {"@type": "NutritionInformation"}
+                recipe_dict["nutrition"]["calories"] = total.xpath("./td[2]")[0].text
+                recipe_dict["nutrition"]["carbohydrateContent"] = total.xpath(
+                    "./td[3]"
+                )[0].text
+                recipe_dict["nutrition"]["proteinContent"] = total.xpath("./td[5]")[
+                    0
+                ].text
+                recipe_dict["nutrition"]["fatContent"] = total.xpath("./td[4]")[0].text
+                recipe_dict["nutrition"]["sugarContent"] = total.xpath("./td[7]")[
+                    0
+                ].text
+                recipe_dict["nutrition"]["sodiumContent"] = total.xpath("./td[6]")[
+                    0
+                ].text
+        except Exception:
+            logger.warning(f"Could not extract meal information from {meal_url}")
+            return None
```
```suggestion
        ingredients = document.xpath('//*[@id="meal-table"]/tbody/tr')
        # No ingredients?
        if (
            len(ingredients) == 1
            and ingredients[0].xpath("./td[1]")[0].text == "\xa0"
        ):
            raise SomeKindOfException("No ingredients?")
        else:
            for ingredient in ingredients:
                recipe_dict["recipeIngredient"].append(
                    ingredient.xpath("./td[1]")[0].text
                )

            total = document.xpath('//*[@id="mealTableTotal"]/tbody/tr')[0]
            recipe_dict["nutrition"] = {"@type": "NutritionInformation"}
            recipe_dict["nutrition"]["calories"] = total.xpath("./td[2]")[0].text
            recipe_dict["nutrition"]["carbohydrateContent"] = total.xpath(
                "./td[3]"
            )[0].text
            recipe_dict["nutrition"]["proteinContent"] = total.xpath("./td[5]")[
                0
            ].text
            recipe_dict["nutrition"]["fatContent"] = total.xpath("./td[4]")[0].text
            recipe_dict["nutrition"]["sugarContent"] = total.xpath("./td[7]")[
                0
            ].text
            recipe_dict["nutrition"]["sodiumContent"] = total.xpath("./td[6]")[
                0
            ].text
```

This is another situation where I think we should just let the exception go uncaught so the caller can possibly handle it.

coddingtonbear reviewed (commented):
Thanks for taking the time to add a feature, @Dnic94!  I think this would be really useful for a lot of people, and apologize for how many comments you're going to see below -- there are quite a few -- but they're mostly centered around a couple categories of suggestions:

1. Types are really helpful for folks who are using libraries like this since they make it clear what to expect when writing code that interacts with the library.  Without them, folks have to read through the source and, in the worst case, run the code manually to see what to expect.  That's really painful for people integrating with a library; so if I could trouble you to add return types to these functions below, I think folks using this would really appreciate it.
2. In general in Python, if you call a method, it either succeeds and gives you whatever it is you're asking for, or it fails and raises an exception.  Here, in a handful of places, you're emitting logging messages to help communicate to the end user what went wrong, but unfortunately the library might not be being used by a person -- it might be wrapped up in some other application or script -- and there won't be anybody watching to read those logging messages.  So, what I've suggested below in a variety of places is to either let an underlying exception go uncaught or to raise an exception of some kind when things fail.  This is both helpful for folks integrating with this library, as well as folks just running code from the shell manually because that exception message will provide enough information to make an actionable bug report.
3. There are a handful of places where you have left commented-out code with the intention of helping future folks implement parts of these features.  I totally get that you're trying to help a future person implement more functionality, but in most cases commented-out code just ends up being noise that gets in the way of folks understanding the actual uncommented logic around them, and in a lot of cases (as APIs drift and other changes occur) they become misleading over time.  I've left suggestions here that you delete those lines to avoid that kind of problem.

Thanks again for taking the time to contribute a new feature, and I hope you forgive me for how many comments and suggestions you're going to see below.

Cheers!

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -804,3 +810,535 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+        serving_size: str = "1 Serving",
+        servingspercontainer: float = 1.0,
+        sharepublic: bool = False,
+    ):
```
Sure, will do this for all added functions within the next commits and give you a heads up ;)

Dnic94 commented:
> Thanks for taking the time to add a feature, @Dnic94! I think this would be really useful for a lot of people, and apologize for how many comments you're going to see below -- there are quite a few -- but they're mostly centered around a couple categories of suggestions:
> 
> 1. Types are really helpful for folks who are using libraries like this since they make it clear what to expect when writing code that interacts with the library.  Without them, folks have to read through the source and, in the worst case, run the code manually to see what to expect.  That's really painful for people integrating with a library; so if I could trouble you to add return types to these functions below, I think folks using this would really appreciate it.
> 2. In general in Python, if you call a method, it either succeeds and gives you whatever it is you're asking for, or it fails and raises an exception.  Here, in a handful of places, you're emitting logging messages to help communicate to the end user what went wrong, but unfortunately the library might not be being used by a person -- it might be wrapped up in some other application or script -- and there won't be anybody watching to read those logging messages.  So, what I've suggested below in a variety of places is to either let an underlying exception go uncaught or to raise an exception of some kind when things fail.  This is both helpful for folks integrating with this library, as well as folks just running code from the shell manually because that exception message will provide enough information to make an actionable bug report.
> 3. There are a handful of places where you have left commented-out code with the intention of helping future folks implement parts of these features.  I totally get that you're trying to help a future person implement more functionality, but in most cases commented-out code just ends up being noise that gets in the way of folks understanding the actual uncommented logic around them, and in a lot of cases (as APIs drift and other changes occur) they become misleading over time.  I've left suggestions here that you delete those lines to avoid that kind of problem.
> 
> Thanks again for taking the time to contribute a new feature, and I hope you forgive me for how many comments and suggestions you're going to see below.
> 
> Cheers!

Hey @coddingtonbear,

First I want to thank you for your great feedback and your time reviewing my code.
I really appreciate it and I am thankful for any chance to learn new things or get a different view on things. I am not a professional developer so this is not my daily business.

I implemented most of your suggestions and implemented some missing things like docstrings and the so often mentioned return types. I also understand why we need to leave some exceptions uncaught which makes total sense from the view of a user of this library.

I respect your personal preferences to keep the code clean and therefore I removed all unnecessary debugging comments and code left overs. 

I left some additional comments above in your commit suggestions. Everything should be fixed, explained or implemented now. 
Again thank you for your time and creating this library. I use it daily!
Have a nice day!

coddingtonbear commented:
Just to set expectations -- I've got a lot going on this week so I'm hoping to have a glance at this again this weekend sometime.  Thanks for having taken the time to polish this up!

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -830,23 +827,23 @@ def set_new_food(
         servingspercontainer: float = 1.0,
         sharepublic: bool = False,
-    ) -> bool:
+    ) -> None:
```
I looked through the method, and it seemed like there were two possibilities: we returned `True` or we raised an exception.  The `bool` return was really just communicating that the operation was successful, and we already know the operation was successful if an exception wasn't raised.

Dnic94 replied:
I understand why you replaced the empty strings to 0. But in most scenarios this is just not the same and wrong. If we want to add a food item to MFP and there is no value for example fiber on the package it does not mean that there is no fiber (0) in it, it is just not provided or unknown. MFP was able to handle the empty post for such micro nutritions and it was correctly shown as “-” on MFP. Now there is a false “0” shown. Unfortunately I don't know what is the best solution for this problem.

To change the return value to None looks fine.

coddingtonbear replied:
So, if it's not clear -- the reason that I changed those is that `""` is not a `float`, which was the type you declared.  Typing tools like `mypy` won't allow you to do things like that.  If we do want to send `""` when a value isn't provided, we probably want to use instead `Optional[float] = None` and translate those `None` values into `""` when sending them to MFP.

Dnic94 replied:
It was clear. However it worked so far, but you are right it was not "clean".
I implemented your suggestion. ;)

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -939,42 +936,39 @@ def set_new_food(
         )
-        if result.status_code == 200:
-            document = lxml.html.document_fromstring(result.content.decode("utf-8"))
+        if not result.ok:
```
I've here switched around the order in which we handle the response so we "fail fast" and thus can remove a layer of indentation in the successful case.  I _believe_ this is equivalent, but please double-check me.

Dnic94 replied:
Looks good :)

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -939,42 +936,39 @@ def set_new_food(
         self,
-        energy: float = None,
-        energy_unit: str = None,
+        energy: float,
```
The linter was complaining to me that the energy value could be `None`  here: https://github.com/coddingtonbear/python-myfitnesspal/pull/121/commits/419c4163da17409be8c52824837abcb53ec8cd05#diff-30f1e1808bae1ffacf52afff3d5e9489e4224b19dfd67f94653af05fcfbf09e0L1095, and when looking through the method it looked like we were always raising an exception if `energy` were not a `float`; so I've here just changed the signature to demand that you provide an energy value and was then able to remove the logic asserting that that value was non-`None`.

Dnic94 replied:
Jep, but now the energy argument is mandatory and this breaks various situations of the function based on the provided arguments. As mentioned in the docstrings the energy argument was not needed if all macro or percentage values for each macro were provided.. Therefore, I only raised an exception for the missing energy value in if-branches where we already checked if the other arguments are missing too. A possible solution would be to set the default value for energy to -1 and replace the None-Checks with energy < 0?

coddingtonbear replied:
I might be missing one of the branches, but it looks like all logic branches here required `energy`, here, no? 

https://github.com/Dnic94/python-myfitnesspal/compare/2f76e2f9fda53a835c79b61dbe5f53fa6447b7e9...80878cf#diff-30f1e1808bae1ffacf52afff3d5e9489e4224b19dfd67f94653af05fcfbf09e0R1020

Dnic94 replied:
Unfortunately, your links seem not to work.

In the else-branch starting line 1072 (commit 419c4163da17409be8c52824837abcb53ec8cd05), the energy value was not needed. I also mentioned it in a comment there, line 1076.

But the value will be corrected anyways if it is not plausible to the passed macros. Now, user can pass “0” for energy or something like that if they want to use this functionality. 
So, an extra hint in the docstrings should be fine and it makes the code much cleaner.

coddingtonbear replied:
Looking at that else branch, it looks like if you were to pass `None` (the default), you'd get a `TypeError` because of this line: https://github.com/coddingtonbear/python-myfitnesspal/commit/419c4163da17409be8c52824837abcb53ec8cd05#diff-30f1e1808bae1ffacf52afff3d5e9489e4224b19dfd67f94653af05fcfbf09e0R1077, no?

Dnic94 replied:
Wow, I thought I tested all cases, but seems like it was not the case...
So, nevermind 😅 🤦‍♂️

coddingtonbear replied:
It happens to all of us!  It's also one of the reasons why unit tests are so popular as a bug-reduction technique.

So, regarding approaches -- I'm not exactly sure what direction to go, but as long as `mypy` isn't complaining about types, I'm pretty alright with any path forward.

Dnic94 replied:
Here is nothing left to adjust. I am fine with your changes and I already adjusted the docstrings in the last commit.
So I think we can close this conversation. 👍

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -996,6 +990,7 @@ def set_new_goal(
 
         # Get User Default Unit Preference
         if energy_unit != "calories" and energy_unit != "kilojoules":
+            assert self.user_metadata
```
Just to keep the linter happy; `self.user_metadata` is technically marked `Optional`, but you're right that you can reasonably rely on it being set.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -1243,13 +1226,15 @@ def get_recipe(self, recipeid: int) -> dict:
 
         # add some required tags to match schema
-        recipe_dict["recipe_instructions"] = []
+        recipe_dict["recipeInstructions"] = ""
```
Looking at the schema for this, I believe this is actually the right key name.

Dnic94 replied:
You are right. 👍

coddingtonbear reviewed (approved):
I made a couple changes over the weekend, and have left some comments to explain a couple tiny changes I've made.  I need to spend a few minutes to put together some documentation covering this, and I hope to get to that maybe this weekend, but this all looks great, @Dnic94! Thanks for posting this PR and working with me on the review process!

Dnic94 commented on myfitnesspal/client.py:
```diff
@@ -1024,69 +1019,56 @@ def set_new_goal(
-                        "No energy value and no macro values provided! - Not able to update goals without a mandotory amount of information."
-                    )
-                else:
-                    if percent_carbohydrates + percent_protein + percent_fat == 100.0:
```
I did not think about that, but interesting. Thanks for the heads up.

coddingtonbear commented on myfitnesspal/client.py:
```diff
@@ -812,3 +818,485 @@ def get_food_item_details(self, mfp_id: int) -> FoodItem:
+            "nutritional_content[calories]": f"{calories}",
+            "nutritional_content[sodium]": f"{sodium or ''}",
+            "nutritional_content[fat]": f"{fat}",
+            "nutritional_content[potassium]": f"{potassium if potassium is not None else ''}",
```
Forgive me for the little tiny correction here -- just to explain, you're relying on the value `None` being "falsy" here, but the value `0` also is.  So if we want to be able to allow folks to specify a value of `0` for these fields (that is: not transform it into an empty string), we have to actually check if the value is `None` instead of just checking if the value is `Falsy`.  All fixed here, though, now.
